[2015.8] Merge forward from 2015.5 to 2015.8

basepi opened the pull request:
[2015.8] Merge forward from 2015.5 to 2015.8

```
Conflicts:
    conf/minion
    salt/client/__init__.py
    salt/states/dockerio.py
    salt/utils/event.py
    tests/unit/utils/event_test.py
```

Commit:
Fixed minion failover to next master on DNS errors.

driskell committed:
Implement full event caching for subscribed tags
Require all multitasking contexts to subscribe to their events so one call to get_event for one tag does not discard events that should be saved for a subsequent call to get_event with another tag.
Use blocking get_event in batching with very small timeout.
Fixes #25998

jacobhammons committed:
Updated windows download links in the docs to https://repo.saltstack.com
Refs #25961

jacobhammons committed:
Merge pull request #26047 from jacobhammons/win-downloads

Updated windows download links in the docs to https://repo.saltstack.com

twangboy committed:
Added python_shell=True, quoted user input

jacobhammons committed:
master_finger configuration docs
switch a script to use https:// instead of http://
Refs #25751

jacobhammons committed:
some small changes

Commit:
Merge pull request #26088 from jacobhammons/master-finger

Master finger

corux committed:
Append/prepend: search for full line with escaped content

cro committed:
Cherry pick master.py groups check

cro committed:
Documentation update for anonymous bind issue.

cro committed:
Cherry pick index.rst change.

martinhoefling committed:
Fixes #26024

loa committed:
Fix dockerio state documentation typo

jfindlay committed:
insert saltenv to render_state args in state.template

Commit:
Merge pull request #26116 from corux/fix-escape-content

file.replace fails if repl string is an invalid regex and append/prepend is used

jfindlay committed:
check type of matches in render_state before iterating

Fixes #26112.

basepi committed:
Merge remote-tracking branch 'upstream/2014.7' into merge-forward-2015.5

Conflicts:
    conf/minion
    doc/conf.py
    salt/modules/file.py

jacobhammons committed:
regenerated man pages

Commit:
Merge pull request #26180 from jfindlay/templ_env

fix processing of state.template

Commit:
Merge pull request #26179 from rallytime/bp-25404

Back-port #25404 to 2015.5

Commit:
Merge pull request #26177 from rallytime/bp-26147

Back-port #26147 to 2015.5

Commit:
Merge pull request #26175 from rallytime/bp-26153

Back-port #26153 to 2015.5

cro committed:
Lint roller

Commit:
Merge pull request #26000 from driskell/fix_discarded_events

Implement full event caching for subscribed tags

jfindlay committed:
Merge pull request #26182 from basepi/merge-forward-2015.5

[2015.5] Merge forward from 2014.7 to 2015.5

jfindlay committed:
Merge pull request #26186 from jacobhammons/man-page-updates

regenerated man pages

Commit:
Merge pull request #26183 from cro/anonldap2

Fix LDAP configuration issue.

twangboy committed:
Merge branch '2015.5' of https://github.com/saltstack/salt into fix_25802

twangboy committed:
Fixed but with multiple packages, was causing tests to fail

Commit:
Merge pull request #26084 from twangboy/fix_25802

Added python_shell=True, quoted user input

basepi committed:
Merge remote-tracking branch 'upstream/2015.5' into merge-forward-2015.8

Conflicts:
    conf/minion
    salt/client/__init__.py
    salt/states/dockerio.py
    salt/utils/event.py
    tests/unit/utils/event_test.py

basepi commented:
@driskell #26000 did not merge forward cleanly at all, because of the new match_function code in salt/utils/event.py. I think I resolved the conflicts correctly, retaining the functionality from both new features, but I would love a review from you, and perhaps @cachedout as well.

basepi committed:
Use list comp instead of filter (py3 compat)

basepi committed:
connect_pub call replaces subscribe call

basepi committed:
Remove stray subscribe

cro commented:
Go Go Jenkins!

driskell commented on salt/utils/event.py:
```diff
@@ -430,18 +468,31 @@ def get_event(self, wait=5, tag='', full=False, use_pending=False,
              - 'regex' : regex search '^' + tag event tags
             Default is opts['event_match_type'] or 'startswith'
 
```
Ah

This breaks the caching. If `_get_event` is called with a different `match_type` than other areas of code are using, the `pending_events` doesn't collect events properly. I guess the issue would only be problematic if the code using `match_type` is running along with other code though.

Fix is in `_check_pending` and `_get_event` when checking for events to cache (based on what was `subscribe`d to), we just need to ignore the match_type and default to startswith. This will **change behaviour** but only where there is code that is trying to use the removed `pending_tags` functionality as well as `match_type`. If there isn't any code using both then it should be fine I think.

If there is other code using `pending_tags` and also `match_type` a larger fix would be needed. First is what is already done which is move from the `pending_tags` way to the `subscribe` way, and change the `subscribe` and `_check_pending` algorithms to allow a custom match_type.

driskell commented:
Do you want me to produce a PR against 2015.8 to do the larger fix anyway?
I guess I could even base it from this PR for you so you could cherry pick into it?

basepi commented:
@driskell That would be great if you could fix it all up. There are currently two tests failing on Ubuntu 14.04:

```
unit.utils.event_test.TestSaltEvent.test_event_many_backlog
unit.utils.event_test.TestSaltEvent.test_event_multiple_clients
```

I am going to skip those tests for now so I can get this merged. Would you mind also unskipping those tests and making sure they pass when you make your fix? You're awesome, thanks!

basepi committed:
Skip failing event tests

@driskell it would be great if you could fix these up when you do your
2015.8 refactor

basepi commented:
@driskell We're working on getting a final RC out for 2015.8.0 and would love to have this caching fixed. Any chance you can get a PR in on 2015.8 today or tomorrow?

driskell commented:
Sorry was offline all weekend but will aim for tomorrow!

basepi commented:
Thanks @driskell! That sounds great.
